src/function/decoding_function.cpp

```diff
@@ -0,0 +1,145 @@
+#include "duckdb/function/decoding_function.hpp"
+#include "duckdb/main/config.hpp"
+
+namespace duckdb {
+
+struct DefaultDecodeMethod {
+	string name;
+	decode_t decode_function;
+	idx_t ratio;
+	idx_t bytes_per_iteration;
+};
+
+void DecodeUTF16(char *encoded_buffer, idx_t &encoded_buffer_current_position, const idx_t encoded_buffer_size,
+                 char *decoded_buffer, idx_t &decoded_buffer_current_position, const idx_t decoded_buffer_size,
+                 char *remaining_bytes_buffer, idx_t &remaining_bytes_size) {
+
+	for (; encoded_buffer_current_position < encoded_buffer_size; encoded_buffer_current_position += 2) {
+		if (decoded_buffer_current_position == decoded_buffer_size) {
+			// We are done
+			return;
+		}
+		const uint16_t ch = static_cast<uint16_t>(
+		    static_cast<unsigned char>(encoded_buffer[encoded_buffer_current_position]) |
+		    (static_cast<unsigned char>(encoded_buffer[encoded_buffer_current_position + 1]) << 8));
+		if (ch >= 0xD800 && ch <= 0xDFFF) {
+			throw InvalidInputException("File is not utf-16 encoded");
+		}
+		if (ch <= 0x007F) {
+			// 1-byte UTF-8 for ASCII characters
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(ch & 0x7F);
+		} else if (ch <= 0x07FF) {
+			// 2-byte UTF-8
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0xC0 | (ch >> 6));
+			if (decoded_buffer_current_position == decoded_buffer_size) {
+				// We are done, but we have to store one byte for the next chunk!
+				encoded_buffer_current_position += 2;
```

Comment:
A lot of the code complexity here seems to stem from the fact that we're allowing arbitrary buffer sizes for the `encoded_buffer` and the `decoded_buffer`. I'm not super opinionated here - the current code is more flexible, but if we enforce that `decoded_buffer` is large enough (i.e. `decoded_buffer_size >= encoded_buffer_size * ratio`) then we can get rid of all of the complexity w.r.t. the `remaining_bytes`.

For the CSV reader at least that seems like an easier option, although perhaps I'm missing future scenarios where the current approach works but such an approach would not..

Reply:
I think this is necessary and a bit hard to remove it, especially if we want to be able to output full buffers to the csv reader to get the max out of parallelization.
I've had a try removing it, but it gets more complicated for when dealing with decoding iterations of  3 bytes.